src/bundle/streaming_reader.zig

```diff
@@ -84,172 +79,84 @@ pub const DecompressingHashReader = struct {
-                    reached_end = true;
-                    break :blk 0;
-                },
-                error.ReadFailed => return error.UnexpectedEndOfStream,
-                error.WriteFailed => unreachable, // fixed buffer writer doesn't fail
-            };
-
-            if (bytes_read == 0) {
-                if (reached_end) {
-                    if (!self.hash_verified) {
-                        var actual_hash: [32]u8 = undefined;
-                        self.hasher.final(&actual_hash);
-                        if (!std.mem.eql(u8, &actual_hash, &self.expected_hash)) {
-                            return error.HashMismatch;
-                        }
-                        self.hash_verified = true;
-                    }
-                    self.finished = true;
-                    break;
-                }
-
-                if (total_read > 0) {
-                    break;
-                }
-                continue;
-            }
-
-            // Update hash with compressed data
-            self.hasher.update(self.in_buffer[0..bytes_read]);
-
-            // Decompress
-            var in_buf = c.ZSTD_inBuffer{ .src = self.in_buffer.ptr, .size = bytes_read, .pos = 0 };
-
-            while (in_buf.pos < in_buf.size) {
-                var out_buf = c.ZSTD_outBuffer{ .dst = self.out_buffer.ptr, .size = self.out_buffer.len, .pos = 0 };
-
-                const result = c.ZSTD_decompressStream(self.dctx, &out_buf, &in_buf);
-                if (c.ZSTD_isError(result) != 0) {
-                    return error.DecompressionFailed;
-                }
-
-                if (out_buf.pos > 0) {
-                    self.out_pos = 0;
-                    self.out_end = out_buf.pos;
-
-                    // Copy what we can to dest
-                    const to_copy = @min(out_buf.pos, dest.len - total_read);
-                    @memcpy(dest[total_read..][0..to_copy], self.out_buffer[0..to_copy]);
-                    self.out_pos = to_copy;
-                    total_read += to_copy;
-
-                    if (total_read == dest.len) {
-                        return total_read;
-                    }
-                }
-
-                // If decompression is complete
-                if (result == 0) {
-                    break;
-                }
-            }
+        if (in_buf.pos < in_buf.size) {
+            self.in_pos = in_buf.pos;
+            self.in_end = in_buf.size;
+        } else {
+            self.in_pos = 0;
+            self.in_end = 0;
         }
 
-        return total_read;
+        r.end = out_buf.pos;
+
+        return 0;
     }
 
+    /// Verify that the hash matches. This should be called after reading is complete.
+    /// If there is remaining data, it will be discarded.
     pub fn verifyComplete(self: *Self) !void {
         // Read any remaining data to ensure we process the entire stream
-        var discard_buffer: [1024]u8 = undefined;
         while (true) {
```

Comment:
Does this still need a `while(true)`?

Reply:
Yes, because discard does not have to discard everything up to the limit. But i found ther is an `discardRemaining` that already does the loop, so I will use that one.